### PR TITLE
[CodeQuality] LogicalToBooleanRector fix left or right argument causing recursion error

### PR DESCRIPTION
**Pull request by Bl00D4NGEL**
refs https://github.com/rectorphp/rector/issues/6667
closes https://github.com/rectorphp/rector/issues/6667

The problem originated from `$node->left` and/or `$node->right` being an instance of `PhpParser\Node\Expr\BinaryOp\LogicalAnd` or `PhpParser\Node\Expr\BinaryOp\LogicalOr` which made rector think it detected a recursion.

Fixing this by recursively calling the refactor method on the left and right properties of the operator until it is not an instance of `LogicalOr` or `LogicalAnd` anymore

### CODE REVIEW

**Comment by Bl00D4NGEL** [comment]
All checks have passed 🎉  Please review @TomasVotruba

**Review by samsonasik — APPROVED** [review]

File: rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php [hunk 1]
```diff
@@ -55,10 +55,31 @@ public function getNodeTypes(): array
      */
     public function refactor(Node $node): ?Node
     {
+        /** @var LogicalAnd|LogicalOr $left */
+        $left = $node->left;
```
Comment (TomasVotruba) [inline rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php]: It seems the left/right can be any `Expr`:

https://github.com/nikic/PHP-Parser/blob/9aebf377fcdf205b2156cb78c0bd6e7b2003f106/lib/PhpParser/Node/Expr/BinaryOp.php#L9-L12

The `@var` assumptions already crashes many Rector rules causing fatal errors. This should be prevent now with `instanceof`

File: rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php [hunk 2]
```diff
@@ -55,10 +55,31 @@ public function getNodeTypes(): array
...
+            /** @var BooleanAnd|BooleanOr $left */
+            $left = $this->refactor($left);
+        }
+
+        /** @var LogicalAnd|LogicalOr $right */
+        $right = $node->right;
+        if ($this->shouldRecursivelyCall($right)) {
+            /** @var BooleanAnd|BooleanOr $right */
+            $right = $this->refactor($right);
+        }
```
Comment (TomasVotruba) [inline rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php]: Recurse call should not be needed, as Rector rule works untill there is a code to be fixed. It's recursive by nature.

File: rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php (hunk 1 above)
Comment (Bl00D4NGEL) [reply rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php]: I ensure that $left and $right are LogicalAnd|LogicalOr with the shouldRecursivelyCall function, however I understand your concern and will look into it later today

Comment (TomasVotruba) [reply rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php]: If so, that should happen in line above.
The `$node` comming to `public function refactor(Node $node): ?Node` can be anything.

File: rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php (hunk 2 above)
Comment (Bl00D4NGEL) [reply rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php]: The original problem stems from the "infinite loop detection" in `\Rector\Core\Validation\InfiniteLoopValidator::process`.

The test fixtures multi_or and multi_and will run into an InfiniteLoopTraversingException.

If I use print_node on the offending node(s) this is the output:
`'(1 === $val1 or 1 === $val2) || 1 === $val3'`
and 
`'(1 === $val1 and 1 === $val2) && 1 === $val3'`

Not sure if this is now a bug in the InfiniteLoopValidator or if it is intended and my recursion is approach is the correct one in this case

File: rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php (hunk 1 above)
Comment (Bl00D4NGEL) [reply rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php]: I changed it so $node->left and $node->right are checked for the correct instances.

Comment (TomasVotruba) [reply rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php]: Thank you

File: rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php [hunk 3]
```diff
@@ -55,10 +55,21 @@ public function getNodeTypes(): array
...
+        $right = $node->right;
+        if ($right instanceof LogicalOr || $right instanceof LogicalAnd) {
+            /** @var BooleanAnd|BooleanOr $right */
+            $right = $this->refactor($right);
         }
 
-        return new BooleanAnd($node->left, $node->right);
+        if ($node instanceof LogicalOr) {
+            return new BooleanOr($left, $right);
+        }
```
Comment (samsonasik) [inline rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php:77]: this should has new line after if end, please rebase, then run:

```
composer fix-cs -- rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php
bin/rector process rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php
```

Comment (Bl00D4NGEL) [reply rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php:77]: Done

File: rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php [hunk 4]
```diff
@@ -55,10 +55,22 @@ public function getNodeTypes(): array
...
+            $right = $this->refactor($right);
+        }
+
         if ($node instanceof LogicalOr) {
-            return new BooleanOr($node->left, $node->right);
+            return new BooleanOr($left, $right);
         }
 
-        return new BooleanAnd($node->left, $node->right);
+        return new BooleanAnd($left, $right);
```
Comment (samsonasik) [inline rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php]: this whole code inside refactor can be moved to separate method, eg: processLogicalToBoolean(), so the recursive method is in the new method.

For clean rebase, I created a blogpost about it: https://samsonasik.wordpress.com/2015/09/16/practical-git-4-rebasing-conflicted-task-branch-against-primary-branch/

**Comment by samsonasik** [comment]
@Bl00D4NGEL there is conflict now, please clean rebase to avoid conflict with unrelated files.

File: rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php (hunk 4 above)
Comment (Bl00D4NGEL) [reply rules/CodeQuality/Rector/LogicalAnd/LogicalToBooleanRector.php]: Moved the code, is this what you wanted?

**Comment by Bl00D4NGEL** [comment]
I seem to be struggling way too much with git.. I hope I did it correctly now?

File: phpstan.neon [hunk 5]
```diff
@@ -552,3 +552,8 @@ parameters:
                 - rules/Php74/Rector/LNumber/AddLiteralSeparatorToNumberRector.php
 
         - '#^Cognitive complexity for "Rector\\CodingStyle\\Naming\\NameRenamer\:\:renameNameNode\(\)" is 13, keep it under 9$#'
+
+        -
+            message: '#The string value "" is repeated 6 times\. Refactor to enum to avoid typos and make clear allowed value#'
+            paths:
+                - packages/BetterPhpDocParser/PhpDocManipulator/PhpDocTagRemover.php #45
```
Comment (TomasVotruba) [inline phpstan.neon]: Thanks, should be fixed in next simplify release :+1:

Comment (samsonasik) [reply phpstan.neon]: removed at https://github.com/rectorphp/rector-src/pull/805

**Comment by TomasVotruba** [comment]
Thank you, looks very clean now :+1:

**Comment by TomasVotruba** [comment]
@Bl00D4NGEL I tried to rebase locally, but failed after 30 mins :D could you re-open new PR with single commit? It would save us some rebase hell

**Comment by samsonasik** [comment]
@Bl00D4NGEL I rebased your commits at PR https://github.com/rectorphp/rector-src/pull/808

**Pull request closed without merging**